### ISSUE TITLE
[Feature] Check for updates if WP-CLI checks for updates

### ISSUE THREAD

**Issue by hirasso** [YahnisElsts/plugin-update-checker#558]
Hi there! 

### The Problem

The WP-CLI command `wp plugin ...` has a few commands where they automatically check for plugin updates, for example:

- `wp plugin status`
- `wp plugin list`

I just noticed that Plugin Update Checker doesn't check for updates if these commands are being executed.

### Proposed Solution

Both these commands call [`wp_update_plugins`](https://github.com/WordPress/WordPress/blob/b2e9235b133e1069bc7dc8f3c719f9b1d8991a7d/wp-includes/update.php#L312) directly if being executed. 

Would it be possible to automatically check for updates in Plugin Update Checker as well, if one of those commands is being executed? I just checked and it seems like in `wp_update_plugins` there is no action being triggered. So it might be hard to achieve updates based on only that function.

### Alternatives considered

Implement a custom wp cli command in userland, something like this:

```php
$bb_checker = PucFactory::buildUpdateChecker(
  "https://bitbucket.org/user/$plugin_slug", // Metadata URL.
  $file_path, // Full path to the main plugin file.
  $plugin_slug, // Plugin slug. Usually it's the same as the name of the directory.
  24*7 // <-- update interval in hours
);

if (defined('WP_CLI') && WP_CLI) {
  \WP_CLI::add_command('bb-updater check', function() use ($bb_updater) {
    $bb_updater->checkForUpdates();
  });
}
```

That works, but it would be amazing to get this out of the box.

**Comment by YahnisElsts** [comment]
`wp_update_plugins` modifies the `update_plugins` transient, which triggers filters that could be used to hook into the function (e.g. `pre_set_site_transient_{$transient}`. However, that transient is also used in other places (in Cron, and potentially on every admin page load), so, for better performance, I would want a way to check if the filter is being called inside WP-CLI, preferably even a way to identify the current WP-CLI command. 

I believe the first part can be covered by checking if the `WP_CLI` constant is defined, but I'm not sure about identifying the running command. Do you happen to know how to do that?

**Comment by hirasso** [comment]
Actually, re-reading your comment:

>`wp_update_plugins` modifies the `update_plugins` transient, which triggers filters that could be used to hook into the function (e.g. `pre_set_site_transient_{$transient}`) 

Maybe we could get away with combining the two? 

1. hook into the appropriate filter
2. check if currently running a WP_CLI command starting with "plugin"
3. run the update checker

**Comment by YahnisElsts** [comment]
It looks like WP-CLI provides a `before_invoke:<command>` [hook](https://make.wordpress.org/cli/handbook/references/internal-api/wp-cli-add-hook/#notes) that can be used to detect when certain (sub-)commands run more directly. I've written a basic implementation that hooks into the relevant `plugin` and `theme` commands. For consistency with `wp_update_plugins()`, this implementation doesn't check for updates every time but just triggers the normal scheduled update check logic.

This could use more testing - give it a try.

**Comment by hirasso** [comment]
Works like a charm! I tested it with two plugins – one managed by WordPress (relevanssi), one managed by plugin update checker (rh-admin-utils). Here is the relevant command output: Before updating puc, rh-admin-utils didn't detect that there was a new version available, after updating puc it detected the new version:

### Before
```
❯ wp plugin list
+---------------------------------------+--------+-----------+------------+
| name                                  | status | update    | version    |
+---------------------------------------+--------+-----------+------------+
| relevanssi                            | active | available | 4.22.0     |
| rh-admin-utils                        | active | none      | 1.9.4      |
+---------------------------------------+--------+-----------+------------+
```

### After
```
❯ wp plugin list
+---------------------------------------+--------+-----------+------------+
| name                                  | status | update    | version    |
+---------------------------------------+--------+-----------+------------+
| relevanssi                            | active | available | 4.22.0     |
| rh-admin-utils                        | active | available | 1.9.4      |
+---------------------------------------+--------+-----------+------------+
```

Also, sorry for keeping you waiting for so long... your help means a lot!

**Comment by YahnisElsts** [comment]
Sounds good. I've made a new release (5.4) that includes this feature, so I'll close the issue now.
